File: algorithms/graph/tarjans-scc/python/tarjans_scc.py
def tarjans_scc(arr: list[int]) -> int:
    n = arr[0]
    m = arr[1]
    adj = [[] for _ in range(n)]
    for i in range(m):
        u = arr[2 + 2 * i]
        v = arr[2 + 2 * i + 1]
        adj[u].append(v)

    index_counter = [0]
    stack = []
    on_stack = [False] * n
    index = [-1] * n
    lowlink = [0] * n
    scc_count = [0]

    def strongconnect(v):
        index[v] = index_counter[0]
        lowlink[v] = index_counter[0]
        index_counter[0] += 1
        stack.append(v)
        on_stack[v] = True

        for w in adj[v]:
            if index[w] == -1:
                strongconnect(w)
                lowlink[v] = min(lowlink[v], lowlink[w])
            elif on_stack[w]:
                lowlink[v] = min(lowlink[v], index[w])

        if lowlink[v] == index[v]:
            scc_count[0] += 1
            while True:
                w = stack.pop()
                on_stack[w] = False
                if w == v:
                    break

    for v in range(n):
        if index[v] == -1:
            strongconnect(v)

    return scc_count[0]

File: algorithms/graph/tarjans-scc/csharp/TarjansScc.cs
using System;
using System.Collections.Generic;

public class TarjansScc
{
    private static int indexCounter;
    private static int sccCount;
    private static int[] disc;
    private static int[] low;
    private static bool[] onStack;
    private static Stack<int> stack;
    private static List<int>[] adj;

    public static int Solve(int[] arr)
    {
        int n = arr[0];
        int m = arr[1];
        adj = new List<int>[n];
        for (int i = 0; i < n; i++)
            adj[i] = new List<int>();
        for (int i = 0; i < m; i++)
        {
            int u = arr[2 + 2 * i];
            int v = arr[2 + 2 * i + 1];
            adj[u].Add(v);
        }

        indexCounter = 0;
        sccCount = 0;
        disc = new int[n];
        low = new int[n];
        onStack = new bool[n];
        stack = new Stack<int>();
        for (int i = 0; i < n; i++) disc[i] = -1;

        for (int v = 0; v < n; v++)
        {
            if (disc[v] == -1)
                Strongconnect(v);
        }

        return sccCount;
    }

    private static void Strongconnect(int v)
    {
        disc[v] = indexCounter;
        low[v] = indexCounter;
        indexCounter++;
        stack.Push(v);
        onStack[v] = true;

        foreach (int w in adj[v])
        {
            if (disc[w] == -1)
            {
                Strongconnect(w);
                low[v] = Math.Min(low[v], low[w]);
            }
            else if (onStack[w])
            {
                low[v] = Math.Min(low[v], disc[w]);
            }
        }

        if (low[v] == disc[v])
        {
            sccCount++;
            while (true)
            {
                int w = stack.Pop();
                onStack[w] = false;
                if (w == v) break;
            }
        }
    }
}
